### ISSUE TITLE
Failed to mount sysroot on reboot for nodes with a 'large' disk

### ISSUE THREAD

**Issue by basvdlei** [coreos/bugs#2485]
# Issue Report #

### Container Linux Version ###

```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1800.5.0
VERSION_ID=1800.5.0
BUILD_ID=2018-07-28-2250
PRETTY_NAME="Container Linux by CoreOS 1800.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```

### Environment ###

VMWare ESXi

### Expected Behavior ###

When rebooting a node with a "large" disk it should be able to mount sysroot.

### Actual Behavior ###

A node with a "large" disk fails to mount sysroot when it's rebooted:

```
systemd[1]: Mounting /sysroot...
EXT4-fs (sda9): ext4_check_descriptors: Block bitmap for group 0 overlaps block group descriptors
EXT4-fs (sda9): group descriptors corrupted!
mount[419]: mount: /sysroot: mount(2) system call failed: Structure needs cleaning.
systemd[1]: sysroot.mount: Mount process exited, code=exited status=32
Failed to mount /sysroot.
```

![sysroot-mount-failed](https://user-images.githubusercontent.com/11786068/43466576-20631710-94e0-11e8-9460-7510a0d68e85.jpg)

### Reproduction Steps ###

  1. Create a CoreOS node with a 3.91TB disk (have not been able to test other sizes yet)
  2. Root filesystem is resized and mounted correctly on the first boot
  3. Reboot the machine
  4. Mount of /sysroot fails during the boot

### Other Information ###

We first observed this issue when a machine with a 3.91TB failed to update from 1745.7.0 to 1800.4.0. Version 1745.7.0 was still able to mount the filesystem while 1800.4.0 gave the error described above.

It looks like some regression was introduced in kernel 4.14.55 with the ext4 changes https://lwn.net/Articles/759535/ and (from what we could gather) this may even be the patch: https://patchwork.ozlabs.org/patch/950668/

All of our machines with smaller disks (<500GB) still boot and reboot correctly.

**Comment by adarshaj** [comment]
This seems to be the fix - https://github.com/torvalds/linux/commit/5012284700775a4e6e3fbe7eac4c543c4874b559 (read the commit msg for details and linked culprit commit at https://github.com/torvalds/linux/commit/8844618d8aa7a9973e7b527d038a2a589665002c -- which is exactly the behavior reported above in the logs). How can we test this kernel?

**Comment by dm0-** [comment]
I've cherry-picked the upcoming ext4 fixes (including the commit you linked) onto the current stable and produced a test image here: http://builds.developer.core-os.net/boards/amd64-usr/1800.5.0%2Bjenkins2-build-1800%2Blocal-1683/coreos_production_image.bin.bz2

Can you confirm that resolves the issue?

**Comment by adarshaj** [comment]
Is there a way to test this without nuking the ROOT labelled partition? (for context, I'm running a node in tectonic cluster with cluo managing upgrades)

**Comment by dm0-** [comment]
If the failure is reproducible by just mounting the root partition, you could try booting the ISO or PXE version and mounting the disk manually.  That way, nothing will be overwritten on persistent storage.

**Comment by basvdlei** [comment]
Even with the test image, I'm still able to reproduce this issue.

I took both a 1800.5.0 image (https://stable.release.core-os.net/amd64-usr/1800.5.0/coreos_production_image.bin.bz2) and the test image of @dm0- above and ran through the following scenario.

* Convert and resize the raw image to a 4TB qcow2 image

```sh
qemu-img convert -p -O qcow2 coreos_production_image.bin coreos_production_image.qcow2
qemu-img resize coreos_production_image.qcow2 4T
```

* Created an booted a KVM VM using this image.
* Let it boot to the login prompt.
* Trigger a reboot.

![kvm-test](https://user-images.githubusercontent.com/11786068/43514409-6c201694-9580-11e8-9024-1f0fe7061288.jpg)

**Comment by basvdlei** [comment]
Did a couple of more tests with different disk sizes (1TB -> 2TB -> 3TB). The 1TB and 2TB cases worked fine. 

The 3TB drive case failed. I also noticed that it displayed an additional log line during the resizing:
```
EXT4-fs (sda9): Converting file system to meta_bg
```

**Comment by bgilbert** [comment]
https://github.com/torvalds/linux/commit/44de022c4382541cebdd6de4465d1f4f465ff1dd is not currently in the stable queue for kernels older than 4.17 because of a trivial patch conflict (see e.g. [4.14](https://www.spinics.net/lists/stable/msg249574.html)).  I've reproduced the issue on 1800.5.0, and confirmed that the combination of https://github.com/torvalds/linux/commit/44de022c4382541cebdd6de4465d1f4f465ff1dd and the other ext4 changes queued for 4.14 fixes the problem.

**Comment by bgilbert** [comment]
Backport [posted to stable@](https://www.spinics.net/lists/stable/msg250414.html).

**Comment by adarshaj** [comment]
So we should wait until a new point release gets to stable channel here: https://coreos.com/releases/ before upgrading, right?

**Comment by bgilbert** [comment]
@adarshaj Correct.  We'll backport the fix to the existing release branches.

**Comment by basvdlei** [comment]
Thanks! Release 1800.6.0 solves this issue for us. We successfully updated our 4TB nodes.

**Comment by adarshaj** [comment]
I can confirm too, all the bare metal instances with 4TB disks have successfully upgraded and harddisks are getting mounted without any issues. Thanks!

I guess this issue can be closed now.

**Comment by bgilbert** [comment]
Fixed in alpha 1855.1.0, beta 1828.3.0, and stable 1800.6.0, and upstream in kernel 4.14.62.  Thanks for reporting.
